packages/lit-html/src/lib/parts.ts

```diff
@@ -0,0 +1,137 @@
+    fragment.append(start!);
+    start = nextNode;
+  }
+  part.__value = nothing;
+  return state;
+};
+
+export const restoreNodePart = (part: NodePart, state: NodePartState) => {
+  // TODO (justinfagnani): make an interal-only interface
+  (part as any).__commitNode((state as NodePartStateInternal).__fragment);
+  part.__value = (state as NodePartStateInternal).__value;
+};
+
+const createMarker = () => document.createComment('');
+
+export const createAndInsertPart = (
+  containerPart: NodePart,
+  refPart?: NodePart
+): NodePart => {
+  const container = containerPart.__startNode.parentNode as Node;
+
+  const endNode =
+    refPart === undefined ? containerPart.__endNode : refPart.__startNode;
+
+  const startNode = container.insertBefore(createMarker(), endNode);
+
+  container.insertBefore(createMarker(), endNode);
+  const newPart = new NodePart(startNode, endNode, containerPart.options);
+  return newPart;
+};
+
+export const updatePart = (part: NodePart, value: unknown) => {
+  part.__setValue(value);
+  return part;
+};
+
+// export const assertNodeMakers = (part: NodePart, v?: unknown) => {
+//   let start = part.__startNode;
+//   let end = part.__endNode;
+
+//   if (start === null) {
+//     throw new Error(`invalid startNode - ${v}`);
+//   }
+//   if (end === null) {
+//     return; // fine?
+//   }
+//   while (start !== end) {
+//     if (start.nextSibling === null) {
+//       throw new Error(`endNode not a sucessor sibling of startNode - ${v}`);
+//     }
+//     start = start.nextSibling
+//   }
+// };
+
+export const insertPartBefore = (
+  containerPart: NodePart,
+  part: NodePart,
+  refPart?: NodePart
+) => {
+  const container = containerPart.__startNode.parentNode!;
+
+  const refNode = refPart ? refPart.__startNode : containerPart.__endNode;
+
+  const endNode = part.__endNode!.nextSibling;
+
+  if (endNode !== refNode) {
+    // assertNodeMakers(part);
+    reparentNodes(container, part.__startNode, endNode, refNode);
+  }
+};
+
+export const removePart = (part: NodePart) => {
+  removeNodes(
+    part.__startNode.parentNode!,
+    part.__startNode,
+    part.__endNode!.nextSibling
+  );
+};
+
+/**
+ * Reparents nodes, starting from `start` (inclusive) to `end` (exclusive),
+ * into another container (could be the same container), before `before`. If
+ * `before` is null, it appends the nodes to the container.
+ */
+export const reparentNodes = (
+  container: Node,
+  start: Node | null,
+  end: Node | null = null,
+  before: Node | null = null
+): void => {
+  while (start !== end) {
+    const n = start!.nextSibling;
+    container.insertBefore(start!, before);
+    start = n;
+  }
+};
+
+/**
+ * Removes nodes, starting from `start` (inclusive) to `end` (exclusive), from
+ * `container`.
+ */
+export const removeNodes = (
+  container: Node,
+  start: Node | null,
+  end: Node | null = null
+): void => {
+  while (start !== end) {
+    const n = start!.nextSibling;
+    container.removeChild(start!);
```

Comment:
Did we want to use `.remove()`?

Reply:
done

Reply:
Wait, if we're using `append` in core, we might as well use `remove`.  Both will need to be in the DOM shim we talked about adding to wcjs platform.